File: 4.14/서지훈_백준_1992_쿼드트리.c
#include <stdio.h>

int video[64][64];

void recursion(int N, int x, int y){
    int check = video[y][x];

	int i, j;
    for(i = y; i < y + N; ++i){
        for(j = x; j < x + N; ++j){
            if(check != video[i][j]){
                printf("(");

                recursion(N / 2, x, y);
                recursion(N / 2, x + N / 2, y);
                recursion(N / 2, x, y + N / 2);
                recursion(N / 2, x + N / 2, y + N / 2);

                printf(")");

                return;
            }
        }
    }

    if (check == 0) printf("0");
    else printf("1");
}

int main(){
	int N;
	scanf("%d", &N);
	
	char str[65];
	
	int i, j;
	for(i = 0; i < N; ++i){
		scanf("%s", &str);
		for(j = 0; j < N; ++j){
			video[i][j] = str[j] - '0';
		}
	}
	
	recursion(N, 0, 0);
}

File: 4.14/서지훈_백준_1992_쿼드트리.kt
import java.io.*

lateinit var video: Array<IntArray>
var bw: BufferedWriter? = null

fun recursion(N: Int, x: Int, y: Int) {
    val check = video[y][x]
    for (i in y until y + N) {
        for (j in x until x + N) {
            if (check != video[i][j]) {
                bw!!.write("(")
                recursion(N / 2, x, y)
                recursion(N / 2, x + N / 2, y)
                recursion(N / 2, x, y + N / 2)
                recursion(N / 2, x + N / 2, y + N / 2)
                bw!!.write(")")
                return
            }
        }
    }
    if (check == 0) bw!!.write("0") else bw!!.write("1")
}

fun main(args: Array<String>) {
    val br = BufferedReader(InputStreamReader(System.`in`))
    bw = BufferedWriter(OutputStreamWriter(System.out))
    val N = br.readLine().toInt()
    video = Array(N) { IntArray(N) }
    for (i in 0 until N) {
        val str = br.readLine()
        for (j in 0 until N) {
            video[i][j] = str[j].code - '0'.code
        }
    }
    recursion(N, 0, 0)
    bw!!.flush()
    bw!!.close()
    br.close()
}